File: typescript/src/marca.ts
/**
 * La marca de TeleHost Maps, lista para montar.
 *
 * POR QUÉ ESTÁ EN EL PAQUETE Y NO EN EL README (2026-08-04): el README decía
 * «mostrá © OpenStreetMap · TeleHost Maps en tu mapa» y eso es una regla que
 * depende de que alguien la lea y la escriba bien. La app de repartidor salió a
 * producción sin una sola mención — ni la marca ni la atribución legal. Lo que
 * el paquete trae puesto se usa; lo que hay que acordarse, no.
 *
 * La atribución LEGAL ya no hace falta pedirla: desde 2026-08-04 viaja dentro
 * del style.json y MapLibre la muestra sola, en web y en nativo. Lo que sigue
 * necesitando código es el WATERMARK visible, que es lo que hay acá.
 */

/** El isotipo, dibujado inline: sin pedirlo a la red y sin depender del origen. */
function isotipoSvg(blanco: boolean): string {
  const relleno = blanco
    ? '#fff'
    : 'url(#thm-g)';
  const defs = blanco
    ? ""
    : '<defs><linearGradient id="thm-g" x1="0" y1="0" x2="1" y2="1">' +
      '<stop offset="0" stop-color="#ED3237"/><stop offset="1" stop-color="#F30"/>' +
      "</linearGradient></defs>";
  // Fracciones tomadas del brand/icon.svg original (viewBox 14000, círculo r=47,5 %).
  const barras = [
    { x: 0.297, y: 0.265, w: 0.413, h: 0.104 },
    { x: 0.278, y: 0.415, w: 0.452, h: 0.176 },
    { x: 0.289, y: 0.635, w: 0.431, h: 0.101 },
  ]
    .map(b => `<rect x="${b.x * 100}" y="${b.y * 100}" width="${b.w * 100}" height="${b.h * 100}" rx="${(b.h * 100) / 2}" fill="${blanco ? "#ED3237" : "#fff"}"/>`)
    .join("");
  return (
    `<svg viewBox="0 0 100 100" width="15" height="15" aria-hidden="true">${defs}` +
    `<circle cx="50" cy="50" r="47.5" fill="${relleno}"/>${barras}</svg>`
  );
}

/** Los valores de la marca, por si tenés que dibujarla vos (React Native, canvas…). */
export const MARCA = {
  texto: "TeleHost Maps",
  enlace: "https://maps.telehost.net",
  rojo: "#ED3237",
  claro: { fondo: "rgba(255,255,255,.8)", tinta: "#1f2937" },
  oscuro: { fondo: "rgba(12,12,15,.55)", tinta: "#f4f4f5" },
  /** El isotipo como SVG inline. `blanco:true` para satélite y temas oscuros. */
  isotipo: isotipoSvg,
} as const;

/**
 * Atribución que agrega la PÁGINA sobre lo que ya dice el estilo.
 * OSM y la marca ya vienen en `sources[].attribution`: repetirlos acá los
 * muestra dos veces (MapLibre deduplica sólo por string exacta).
 * Overture sí va, y sólo si mostrás resultados de búsqueda: son sus datos.
 */
export const ATRIBUCION_OVERTURE =
  '<a href="https://overturemaps.org" target="_blank" rel="noopener">Overture Maps</a>';

export interface OpcionesMarca {
  /** Tema oscuro: para satélite, híbrido y dark-matter. */
  oscuro?: boolean;
}

/** Lo mínimo de un control de MapLibre, sin depender del tipo de la librería. */
interface ControlMapa {
  onAdd(): HTMLElement;
  onRemove(): void;
  getDefaultPosition?(): string;
}

/**
 * Watermark de TeleHost Maps para **maplibre-gl en web**.
 *
 * ```ts
 * import { controlDeMarca } from "@telehostnet/maps";
 * const marca = controlDeMarca();
 * map.addControl(marca, "bottom-left");
 * marca.tema(true);            // al pasar a satélite
 * ```
 *
 * En React Native el mapa es nativo y no hay DOM: usá `MARCA` para armar la
 * misma píldora con un `View` (mismos colores, mismo texto, mismo enlace).
 */
export function controlDeMarca(opc: OpcionesMarca = {}): ControlMapa & { tema(oscuro: boolean): void } {
  let oscuro = !!opc.oscuro;
  let raiz: HTMLElement | null = null;

  const pintar = () => {
    if (!raiz) return;
    const t = oscuro ? MARCA.oscuro : MARCA.claro;
    raiz.innerHTML =
      `<a href="${MARCA.enlace}" target="_blank" rel="noopener" title="${MARCA.texto}" ` +
      `style="display:flex;align-items:center;gap:6px;padding:4px 9px;border-radius:9px;` +
      `text-decoration:none;background:${t.fondo};color:${t.tinta};` +
      `font:600 12px/1 system-ui,-apple-system,'Segoe UI',Roboto,sans-serif;` +
      `box-shadow:0 1px 4px rgba(15,23,42,.18);backdrop-filter:blur(4px);-webkit-backdrop-filter:blur(4px)">` +
      MARCA.isotipo(oscuro) +
      `<span>TeleHost&nbsp;<b style="font-weight:800">Maps</b></span></a>`;
  };

  return {
    onAdd() {
      raiz = document.createElement("div");
      raiz.className = "maplibregl-ctrl thm-marca";
      pintar();
      return raiz;
    },
    onRemove() {
      raiz?.parentNode?.removeChild(raiz);
      raiz = null;
    },
    getDefaultPosition: () => "bottom-left",
    /** Cambia el tema en vivo, al alternar de basemap. */
    tema(o: boolean) {
      oscuro = !!o;
      pintar();
    },
  };
}

File: flutter/lib/src/marca.dart
/// La marca de TeleHost Maps, lista para poner encima del mapa.
///
/// POR QUÉ ESTÁ EN EL PAQUETE Y NO EN EL README (2026-08-04): el README decía
/// «mostrá © OpenStreetMap · TeleHost Maps en tu mapa» y eso es una regla que
/// depende de que alguien la lea y la escriba bien. La app de repartidor salió a
/// producción sin una sola mención — ni la marca ni la atribución legal. Lo que
/// el paquete trae puesto se usa; lo que hay que acordarse, no.
///
/// La atribución LEGAL ya no hay que dibujarla: desde 2026-08-04 viaja dentro
/// del style.json y MapLibre nativo la muestra sola. Lo único que hay que hacer
/// es no esconder el botón de atribución del mapa: es la licencia ODbL de
/// OpenStreetMap, no una preferencia nuestra.
library;

import 'dart:math' as math;
import 'package:flutter/widgets.dart';

/// El isotipo: círculo con degradado y tres barras. Se DIBUJA, no se descarga:
/// un repartidor sin señal tiene que ver la marca igual, y así el paquete no
/// arrastra `flutter_svg` ni depende de que el dominio resuelva.
class _Isotipo extends CustomPainter {
  const _Isotipo({required this.blanco});
  final bool blanco;

  // Fracciones tomadas del brand/icon.svg original (viewBox 14000).
  static const _barras = <List<double>>[
    [0.297, 0.265, 0.413, 0.104],
    [0.278, 0.415, 0.452, 0.176],
    [0.289, 0.635, 0.431, 0.101],
  ];

  @override
  void paint(Canvas lienzo, Size s) {
    final centro = Offset(s.width / 2, s.height / 2);
    final radio = math.min(s.width, s.height) * 0.475;
    final disco = Paint()..isAntiAlias = true;
    if (blanco) {
      disco.color = const Color(0xFFFFFFFF);
    } else {
      disco.shader = const LinearGradient(
        begin: Alignment.topLeft,
        end: Alignment.bottomRight,
        colors: [Color(0xFFED3237), Color(0xFFFF3300)],
      ).createShader(Rect.fromCircle(center: centro, radius: radio));
    }
    lienzo.drawCircle(centro, radio, disco);

    final barra = Paint()
      ..isAntiAlias = true
      ..color = blanco ? const Color(0xFFED3237) : const Color(0xFFFFFFFF);
    for (final b in _barras) {
      final r = Rect.fromLTWH(b[0] * s.width, b[1] * s.height, b[2] * s.width, b[3] * s.height);
      lienzo.drawRRect(RRect.fromRectAndRadius(r, Radius.circular(r.height / 2)), barra);
    }
  }

  @override
  bool shouldRepaint(_Isotipo viejo) => viejo.blanco != blanco;
}

/// Watermark de TeleHost Maps: isotipo + «TeleHost **Maps**» en una píldora
/// translúcida, como la de Mapbox.
///
/// Va **abajo a la izquierda** (la atribución del mapa va abajo a la derecha:
/// no se cruzan), y se pone encima del mapa con un `Stack`:
///
/// ```dart
/// Stack(children: [
///   MapLibreMap(styleString: maps.urlEstilo(), /* … */),
///   const Positioned(left: 10, bottom: 10, child: MarcaTeleHost()),
/// ])
/// ```
///
/// Sobre satélite o híbrido, `MarcaTeleHost(oscuro: true)`: una píldora blanca
/// sobre imagen satelital no se lee.
///
/// El wordmark es TEXTO, no el logotipo vectorial: a 15 px de alto ese logotipo
/// se destruye y el texto del sistema se lee nítido.
class MarcaTeleHost extends StatelessWidget {
  const MarcaTeleHost({super.key, this.oscuro = false, this.alTocar});

  /// Tema oscuro, para satélite / híbrido / dark-matter.
  final bool oscuro;

  /// Qué hacer al tocarla. Lo normal es abrir https://maps.telehost.net —
  /// se deja al que integra porque abrir una URL necesita `url_launcher`, y
  /// este paquete no arrastra dependencias.
  final VoidCallback? alTocar;

  /// El enlace al que debería llevar la marca.
  static const enlace = 'https://maps.telehost.net';

  @override
  Widget build(BuildContext context) {
    final tinta = oscuro ? const Color(0xFFF4F4F5) : const Color(0xFF1F2937);
    final pildora = Container(
      padding: const EdgeInsets.symmetric(horizontal: 9, vertical: 4),
      decoration: BoxDecoration(
        color: oscuro ? const Color(0x8C0C0C0F) : const Color(0xCCFFFFFF),
        borderRadius: BorderRadius.circular(9),
        boxShadow: [
          BoxShadow(
            color: oscuro ? const Color(0x59000000) : const Color(0x2E0F172A),
            blurRadius: 4,
            offset: const Offset(0, 1),
          ),
        ],
      ),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          SizedBox(
            width: 15,
            height: 15,
            child: CustomPaint(painter: _Isotipo(blanco: oscuro)),
          ),
          const SizedBox(width: 6),
          Text.rich(
            TextSpan(children: [
              const TextSpan(text: 'TeleHost '),
              TextSpan(text: 'Maps', style: TextStyle(fontWeight: FontWeight.w800, color: tinta)),
            ]),
            textDirection: TextDirection.ltr,
            style: TextStyle(fontSize: 12, fontWeight: FontWeight.w600, color: tinta, height: 1),
          ),
        ],
      ),
    );
    if (alTocar == null) return pildora;
    return GestureDetector(onTap: alTocar, child: pildora);
  }
}
